internal/pkg/instrumentation/bpf/net/http/client/bpf/probe.bpf.c

```diff
@@ -267,4 +174,54 @@ int uprobe_Transport_roundTrip_Returns(struct pt_regs *ctx) {
 
     bpf_map_delete_elem(&http_events, &key);
     return 0;
+}
+
+SEC("uprobe/header_writeSubset")
+int uprobe_writeSubset(struct pt_regs *ctx) {
+    u64 headers_pos = 1;
+    void *headers_ptr = get_argument(ctx, headers_pos);
+
+    u64 io_writer_pos = 3;
+    void *io_writer_ptr = get_argument(ctx, io_writer_pos);
+
+    void **key_ptr = bpf_map_lookup_elem(&http_headers, &headers_ptr);
+    if (key_ptr) {
+        void *key = *key_ptr;
+
+        struct http_request_t *http_req_span = bpf_map_lookup_elem(&http_events, &key);
+        if (http_req_span) {
+            char tp[W3C_VAL_LENGTH];
+            span_context_to_w3c_string(&http_req_span->sc, tp);
+
+            void *buf_ptr = 0;
+            bpf_probe_read(&buf_ptr, sizeof(buf_ptr), (void *)(io_writer_ptr + io_writer_buf_ptr_pos)); // grab buf ptr
+            if (!buf_ptr) {
+                goto done;
+            }
+
+            s64 size = 0;
+            bpf_probe_read(&size, sizeof(s64), (void *)(io_writer_ptr + io_writer_buf_ptr_pos + 8)); // grab size
+
+            s64 len = 0;
+            bpf_probe_read(&len, sizeof(s64), (void *)(io_writer_ptr + io_writer_n_pos)); // grab len
+
+#ifndef NO_HEADER_PROPAGATION
+            if (len < (size - W3C_VAL_LENGTH - W3C_KEY_LENGTH - 4)) { // 4 = strlen(":_") + strlen("\r\n")
+                char key[W3C_KEY_LENGTH + 2] = "Traceparent: ";
+                char end[2] = "\r\n";
+                bpf_probe_write_user(buf_ptr + (len & 0x0ffff), key, sizeof(key));
+                len += W3C_KEY_LENGTH + 2;
+                bpf_probe_write_user(buf_ptr + (len & 0x0ffff), tp, sizeof(tp));
+                len += W3C_VAL_LENGTH;
+                bpf_probe_write_user(buf_ptr + (len & 0x0ffff), end, sizeof(end));
+                len += 2;
+                bpf_probe_write_user((void *)(io_writer_ptr + io_writer_n_pos), &len, sizeof(len));
```

Comment:
Should these 4 write operation be "atomic"? i.e. if one of them fails, we might create an invalid header?

Reply:
Done thanks for the suggestion!

Reply:
Actually, I meant that if it is possible to replace these 4 write operations with a single `bpf_probe_write_user` it will be better.

Reply:
Sorry, I misunderstood. I've changed the code now to use a single char buffer which combines the the 'traceparent' key and value, and I call a single bpf_probe_write_user to store the full buffer. Updating the 'N' (the new length of the buffer) must be done through separate write, however if we fail the first write it won't affect the second. And if the second fails (updating the n), we simply wouldn't get a traceparent in the buffer.